File: BOJ/9291.cpp
#include <bits/stdc++.h>
using namespace std;

int sudoku[9][9];
int check[10];

void resetSudoku() { // test case마다 주어지는 값이 다르므로 test case마다 변수를 초기화 해야한다.
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			sudoku[i][j] = 0;
		}
	}
}

void resetCheck() { // check배열 초기화
	for (int i = 1; i <= 9; i++) check[i] = 0;
}

bool isCorrect() {
	for (int i = 1; i <= 9; i++) {
		if (check[i] != 1) return false;
	}

	return true;
}

string checkSudoku() {
	bool correct = true;

	// 가로줄 체크
	for (int y = 0; y < 9; y++) {
		resetCheck(); // 가로줄을 기준으로 규칙이 맞는지 체크하기 위해 가로줄을 한 번 탐색할 때마다 check배열을 초기화
		for (int x = 0; x < 9; x++) {
			check[sudoku[y][x]]++;
		}
		if (!isCorrect()) return "INCORRECT";
	}

	// 세로줄 체크
	for (int x = 0; x < 9; x++) {
		resetCheck(); // 세로줄을 기준으로 규칙이 맞는지 체크하기 위해 세로줄을 한 번 탐색할 때마다 check배열을 초기화
		for (int y = 0; y < 9; y++) {
			check[sudoku[y][x]]++;
		}
		if (!isCorrect()) return "INCORRECT";
	}

	// 3x3작은 사각형 체크
	for (int dy = 0; dy < 3; dy++) {
		for (int dx = 0; dx < 3; dx++) {
			resetCheck(); // 사각형을 한 번 확인할 때마다 check배열을 초기화 해주어야 한다.
			for (int y = 0; y < 3; y++) {
				for (int x = 0; x < 3; x++) {
					check[sudoku[y + (dy * 3)][x + (dx * 3)]]++;
				}
			}
			if (!isCorrect()) return "INCORRECT";
		}
	}

	return "CORRECT";
}

int main() {
	int tc;
	cin >> tc;
	for (int i = 1; i <= tc; i++) {
		resetSudoku();
		resetCheck();
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				cin >> sudoku[i][j];
			}
		}

		cout << "Case " << i << ": " << checkSudoku() << '\n';
	}

	return 0;
}

File: BOJ/python/9291.py
import sys

input = sys.stdin.readline


def isCorrect(check: list) -> bool:
    """
    주어진 스도쿠가 규칙에 맞는지 확인.

    Args:
        check (list): 1~9까지 한 번씩만 등장했는지 check하기 위한 list

    Returns:
        bool: 규칙에 맞다면 True, 어긋나면 False 반환
    """
    for i in range(1, 10):
        if check[i] != 1:
            return False
    return True


def sudokuCheck(sudoku: list, check: list) -> str:
    # 가로줄 체크
    for y in range(9):
        check = [0 for _ in range(10)]
        for x in range(9):
            check[sudoku[y][x]] += 1
        if not isCorrect(check):
            return "INCORRECT"

    # 세로줄 체크
    for x in range(9):
        check = [0 for _ in range(10)]
        for y in range(9):
            check[sudoku[y][x]] += 1
        if not isCorrect(check):
            return "INCORRECT"

    # 3x3 크기의 작은 사각형 체크
    for dy in range(3):
        for dx in range(3):
            check = [0 for _ in range(10)]
            for y in range(3):
                for x in range(3):
                    check[sudoku[y + (dy * 3)][x + (dx * 3)]] += 1
            if not isCorrect(check):
                return "INCORRECT"
    return "CORRECT"


tc = int(input())
for i in range(1, tc + 1):
    sudoku = []
    sudoku = [list(map(int, input().rstrip("\n").split(" "))) for y in range(9)]
    check = [0 for _ in range(10)]
    print(f"Case {i}: {sudokuCheck(sudoku, check)}")
    input()  # test case마다 개행문자로 구분 되어 있으므로 input을 한번해준다.
